src/WebJobs.Script.WebHost/Security/BlobStorageSecretsRepository.cs

```diff
@@ -0,0 +1,163 @@
+        {
+            return secretsType == ScriptSecretsType.Host
+                ? _hostSecretsBlobPath
+                : string.Format("{0}/{1}", _secretsBlobPath, string.Format(CultureInfo.InvariantCulture, "{0}.json", functionName.ToLowerInvariant()));
+        }
+
+        private string GetSecretsFilePath(ScriptSecretsType secretsType, string functionName = null)
+        {
+            return secretsType == ScriptSecretsType.Host
+                ? _hostSecretsMarkerPath
+                : GetFunctionSecretsFilePath(functionName);
+        }
+
+        private string GetFunctionSecretsFilePath(string functionName)
+        {
+            string secretFileName = string.Format(CultureInfo.InvariantCulture, "{0}.json", functionName.ToLowerInvariant());
+            return Path.Combine(_secretsFileMarkersPath, secretFileName);
+        }
+
+        private void OnChanged(object sender, FileSystemEventArgs e)
+        {
+            var changeHandler = SecretsChanged;
+            if (changeHandler != null)
+            {
+                var args = new SecretsChangedEventArgs { SecretsType = ScriptSecretsType.Host };
+
+                if (string.Compare(Path.GetFileName(e.FullPath), ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) != 0)
+                {
+                    args.SecretsType = ScriptSecretsType.Function;
+                    args.Name = Path.GetFileNameWithoutExtension(e.FullPath).ToLowerInvariant();
+                }
+
+                changeHandler(this, args);
+            }
+        }
+
+        public async Task<string> ReadAsync(ScriptSecretsType type, string functionName)
+        {
+            string filePath = GetSecretsFilePath(type, functionName);
+            string blobPath = GetSecretsBlobFilePath(type, functionName);
+            string secretsContent = null;
+
+            // load the secrets from blob
+            if (_blobContainer.GetBlockBlobReference(blobPath).Exists())
+            {
+                CloudBlockBlob secretBlob = _blobContainer.GetBlockBlobReference(blobPath);
+                secretsContent = await secretBlob.DownloadTextAsync();
+            }
+
+            return secretsContent;
+        }
+
+        public async Task WriteAsync(ScriptSecretsType type, string functionName, string secretsContent)
+        {
+            if (secretsContent == null)
+            {
+                throw new ArgumentNullException(nameof(secretsContent));
+            }
+
+            // get blob file path and write secret to blob file, update file marker
+            string blobPath = GetSecretsBlobFilePath(type, functionName);            
+            CloudBlockBlob secretBlob = _blobContainer.GetBlockBlobReference(blobPath);
+
+            using (StreamWriter writer = new StreamWriter(secretBlob.OpenWrite()))
+            {                
+                writer.Write(secretsContent);
```

Comment:
Use `WriteAsync`  and `OpenWriteAsync`. Alternatively, use  one of the asynchronous write methods on `CloudBlockBlob` (e.g. `UploadTextAsync`) here.